components/tikv_util/src/sys/cgroup.rs

```diff
@@ -1,520 +1,322 @@
+//   * content: a number or "-1" for no limitation
+// ### cpu subsystem, cpu quota
+// v2:
+//   * path: /sys/fs/cgroup/<path-to-group>/cpu.max
+//   * content: 2 numbers seperated with a blank, "max" indicates no limitation.
+// v1:
+//   * path: /sys/fs/cgroup/cpu/<path-to-group>/cpu.cfs_quota_us
+//   * content: a number
+//   * path: /sys/fs/cgroup/cpu/<path-to-group>/cpu.cfs_period_us
+//   * content: a number
+// ### cpuset subsystem, cpus
+// v2:
+//   * path: /sys/fs/cgroup/<path-to-group>/cpuset.cpus
+// v1:
+//   * path: /sys/fs/cgroup/cpuset/<path-to-group>/cpuset.cpus
 
-    Count,
-}
-
-#[derive(Debug, PartialEq)]
-pub struct MountPoint {
-    pub mount_id: i32,
-    pub parent_id: i32,
-    pub device_id: String,
-    pub root: String,
-    pub mount_point: String,
-    pub options: Vec<String>,
-    pub optional_fields: Vec<String>,
-    pub fs_type: String,
-    pub mount_source: String,
-    pub super_options: Vec<String>,
+pub struct CGroupSys {
+    cgroups: HashMap<String, String>,
+    is_v2: bool,
 }
 
-impl MountPoint {
-    // Converts an absolute path inside the *MountPoint's file system to
-    // the host file system path in the mount namespace the *MountPoint belongs to.
-    pub fn translate(&self, abs_path: &str) -> Result<String, Error> {
-        let abs = Path::new(abs_path);
-        let rel = abs.strip_prefix(&self.root)?;
-        let path = Path::new(&self.mount_point).join(rel);
-        if let Some(path) = path.to_str() {
-            Ok(String::from(path))
+impl CGroupSys {
+    pub fn new() -> Self {
+        let lines = read_to_string(&format!("/proc/{}/cgroup", process::id())).unwrap();
+        let is_v2 = is_cgroup2_unified_mode();
+        let cgroups = if !is_v2 {
+            parse_proc_cgroup_v1(&lines)
         } else {
-            Err(box_err!("empty path"))
-        }
-    }
-}
-
-pub fn parse_mount_point_from_line(line: &str) -> Result<MountPoint, String> {
-    let fields: Vec<String> = line.split(MOUNTINFO_SEP).map(|s| s.to_string()).collect();
-    if fields.len() < MountInfoFieldPart1::Count as usize + MountInfoFieldPart2::Count as usize {
-        return Err("mount point format invalid".to_string());
+            parse_proc_cgroup_v2(&lines)
+        };
+        CGroupSys { cgroups, is_v2 }
     }
 
-    let mut sep_pos = MountInfoFieldPart1::OptionalFields as usize;
-    let mut found_sep = false;
-    for field in &fields[MountInfoFieldPart1::OptionalFields as usize..] {
-        if field == OPTIONAL_FIELDS_SEP {
-            found_sep = true;
-            break;
-        }
-        sep_pos += 1;
-    }
-    if !found_sep {
-        return Err(
-            "mount point format invalid, doesn't found optional field separator".to_string(),
-        );
+    /// -1 means no limit.
+    pub fn memory_limit_in_bytes(&self) -> i64 {
+        let path = if self.is_v2 {
+            let group = self.cgroups.get("").unwrap();
```

Comment:
How about log a warning instead? I worry if TiKV can't parse some special system and then cause panic.

Reply:
According to the spec:
> “/proc/$PID/cgroup” lists a process’s cgroup membership. If legacy cgroup is in use in the system, this file may contain multiple lines, one for each hierarchy. The entry for cgroup v2 is always in the format “0::$PATH”
For cgroup-v2 there is only one subsystem named `""`, so here just `unwrap` is OK.

I will attach the spec link to the PR and comment here.